### DISCUSSION TITLE
How to match reports to messages in sendAll()?

### DISCUSSION THREAD

**Discussion by halaei** [kreait/firebase-php#495] [Q&A]
Given the following code from documentations
```php
$messages = [
    // Up to 500 items, either objects implementing Kreait\Firebase\Messaging\Message
    // or arrays that can be used to create valid to Kreait\Firebase\Messaging\Cloudmessage instances
];

/** @var Kreait\Firebase\Messaging\MulticastSendReport $sendReport **/
$sendReport = $messaging->sendAll($messages);
```

After `sendAll()` returns, I am wondering how can I match `$sendReport` items to items of `$messages`. Does the following always work?

```php
foreach ($sendReport->getItems() as $index => $report) {
    $message = $messages[$index];
    // process the ($message, $report) pair
}
```

If so, it would be great to add this to the documentations. If not, then I guess it is a bug needed to be addressed.

Thanks

**Comment by halaei** [comment]
I think it doesn't work in the following cases:
1. `$messages` array is an associated array.
2. any of the conditional `continue` lines in `fromRequestsAndResponses()` function get executed. I am not sure when the conditions are met.

#### Some side notes
I think the time complexity of `fromRequestsAndResponses()` is quadratic, although linear time implementations might be possible.
```php
public static function fromRequestsAndResponses(Requests $requests, Responses $responses)
{
    ...
    // current implementation:
    foreach ($responses as $response) {
        $matchingRequest = $requests->findBy(...); // uses array_filter() and although the name suggests it is a query operation it changes the requests by array_shift too. 
    }
    // linear-time implementation: requests and responses should be made in a way such that requests[i] be always matched with response[i] 
       foreach ($responses as $key => $response) {
        $matchingRequest = $requests[$key];
    }
   ...
    return self::withItems($reports);
}
```
```php
    /**
     * @param SendReport[] $items
     */
    public static function withItems(array $items): self
    {
        $report = new self();
        // current implementation (quadratic time unnecessarily complicated)
        foreach ($items as $item) {
            $report = $report->withAdded($item);
        }
        // simple implementation:
        $report->items = $items;
        return $report;
    }
```

**Comment by jeromegamez** [comment]
@halaei Sorry for the late reply! Could you please have look at #518?

The array index based process that you initially asked about would only work if FCM processes the requests in the same order as we send them. That's why I went with a mapping based on the `Content-ID` header and attached the sent message to each individual `SendReport`.

I hope this is a solution that works for you, let me know if you see further improvements or have other remarks 🙏. Thanks!
